### PR TITLE
Dashboards: Fix folder picker not showing correct results when typing too fast

### PR DESCRIPTION
**Pull request by joshhunt**
**What this PR does / why we need it**:

**tl;dr** don't use debounce with async functions if you care about the return result.

The Folder Picker component in the Dashboard Save component will not show the results for the current input if you type your search in quickly. You can reproduce this on [play.grafana.com](https://play.grafana.org/) by saving a new dashboard and in the folder picker _quickly_ type `iss`. The top result _should_ be "Replicating Issues", but its not. If you slowly type `iss`, then "Replicating Issues" will be the top issue as expected. The user's last input will essentially be discarded.

This issue is caused by debouncing `loadOptions` while trying to use its async return value. While the inner function will be called at the trailing end (for the user's last input), it will happen after the debounced function resolves, and because a function cannot return multiple times, the value is essentially discarded and never available for react-select. 

Here's a simplified timeline for the user searching for `Fo`(der)

 - User types `F`, search query is `F`
   - `debouncedFn(F)` is called by react-select. This triggers the 'leading' call
       - `fn(F)` is called and returns `searchResults<F>`
       - `debouncedFn(F)` returns `searchResults<F>`
   - react-select is showing results for `F`
 - User types `o` within the debounce period, search query is `Fo`
    - `debouncedFn(Fo)` is called by react-select
       - `fn` is _not_ called because we're within the debounce period
       - `debouncedFn(Fo)` returns `searchResults<F>`, from the previous call
   - react-select is showing results for `F`
 - User pauses input and causes the debounce period to expire
    - `fn(Fo)` is called by _lodash.debounce_. This is the 'trailing' call
       - `fn(Fo)` returns the results `searchResults<Fo>`
       - because all calls to `debouncedFn()` have already resolved, the return value from `fn(Fo)` cannot be returned from anywhere.
   - react-select is showing results for `F`

The solution is to not rely on the return value on the debounced function and instead pass a callback to `fn` so it can pass its value to react-select directly.

### CODE REVIEW

File: public/app/core/components/Select/FolderPicker.tsx [hunk 1]
```diff
@@ -82,7 +82,13 @@ export class FolderPicker extends PureComponent<Props, State> {
     await this.loadInitialValue();
   };
 
-  getOptions = async (query: string) => {
+  // when debouncing, we must use the callback form of react-select's loadOptions so we don't
+  // drop results for user input
+  loadOptions = (query: string, callback: LoadOptionsCallback<number>): void => {
+    this.searchFolders(query).then(callback);
+  };
```
Comment (ashharrison90) [inline public/app/core/components/Select/FolderPicker.tsx:89]: seems weird to mix promise styles. could we `async`/`await` here?

**Review by ashharrison90 — APPROVED** [review]
looks good! minor nit

File: public/app/core/components/Select/FolderPicker.tsx (hunk 1 above)
Comment (joshhunt) [reply public/app/core/components/Select/FolderPicker.tsx:89]: No, we specifically cannot - that's what this whole PR is about 😅 

`await` triggers a the limitation that causes results to be dropped.

Comment (ashharrison90) [reply public/app/core/components/Select/FolderPicker.tsx:89]: @joshhunt are you sure? i don't see how this is any different functionally. it's just the promise syntax that's changed: 
```suggestion
  async loadOptions = (query: string, callback: LoadOptionsCallback<number>): void => {
    const options = await this.searchFolders(query);
    callback(options);
  };
```

Comment (joshhunt) [reply public/app/core/components/Select/FolderPicker.tsx:89]: `async` functions return Promises. [An async function cannot return `void`](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwghgGwQIzgYwNZQLxQBQAmcwcAXFAK4B2GVA9gO5UCUOAfFAG50CWBAUPzQI4AZ1FQAEhCR0oAb35RlUBHTgEA8mGA86VCbjEgqafAEcKEAE4hyo4NZ5UA5gBooaRCnQZy8JFRMZnJuPnYFJRVotH0HKGsIUTAcKDgGOB5gKAAzCGA0AAs8SxsQZgBuKOjlL0DfPETkyuqoAF8qtqA), and _returning_ a value from a debounced function is what causes the iffy behaviour.
